fog/sql_recovery_db/src/lib.rs

```diff
@@ -2289,6 +2385,100 @@ mod tests {
+
+        let (block1, records1) = random_block(&mut rng, 122, 10);
+        db.add_block_data(&invoc_id1, &block1, 0, &records1)
+            .unwrap();
+
+        let (block2, records2) = random_block(&mut rng, 123, 10);
+        db.add_block_data(&invoc_id2, &block2, 0, &records2)
+            .unwrap();
+
+        // Get tx outs for a key we're not aware of or a block id we're not aware of
+        // should return empty vec
+        let batch_result = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, 124, 2)
+            .unwrap();
+        assert_eq!(batch_result.len(), 0);
+
+        let batch_result = db
+            .get_tx_outs_by_block_range_and_key(
+                CompressedRistrettoPublic::from_random(&mut rng),
+                123,
+                2,
+            )
+            .unwrap();
+        assert_eq!(batch_result.len(), 0);
+
+        // Getting tx outs in a batch should work as expected when requesting things
+        // that exist
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 1)
+            .unwrap();
+        assert_eq!(batch_results.len(), 1);
+        assert_eq!(batch_results[0], records1);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 1)
+            .unwrap();
+        assert_eq!(batch_results.len(), 1);
+        assert_eq!(batch_results[0], records2);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 2)
+            .unwrap();
+        assert_eq!(batch_results.len(), 2);
+        assert_eq!(batch_results[0], records1);
+        assert_eq!(batch_results[1], records2);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 2)
+            .unwrap();
+        assert_eq!(batch_results.len(), 1);
+        assert_eq!(batch_results[0], records2);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 3)
+            .unwrap();
+        assert_eq!(batch_results.len(), 2);
+        assert_eq!(batch_results[0], records1);
+        assert_eq!(batch_results[1], records2);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 3)
+            .unwrap();
+        assert_eq!(batch_results.len(), 1);
+        assert_eq!(batch_results[0], records2);
+
+        // When there is a gap in the data, the gap should suppress any further results
+        // even if there are hits later in the range.
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index - 1, 2)
+            .unwrap();
+        assert_eq!(batch_results.len(), 0);
+
+        let batch_results = db
+            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index - 2, 3)
+            .unwrap();
+        assert_eq!(batch_results.len(), 0);
```

Comment:
Optional/Equivalent: Assert on the entire Vec
```suggestion
        assert_eq!(batch_result, vec![]);

        let batch_result = db
            .get_tx_outs_by_block_range_and_key(
                CompressedRistrettoPublic::from_random(&mut rng),
                123,
                2,
            )
            .unwrap();
        assert_eq!(batch_result, vec![]);

        // Getting tx outs in a batch should work as expected when requesting things
        // that exist
        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 1)
            .unwrap();
        assert_eq!(batch_result, vec![records1]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 1)
            .unwrap();
        assert_eq!(batch_result, vec![records2]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 2)
            .unwrap();
        assert_eq!(batch_result, vec![records1, records2]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 2)
            .unwrap();
        assert_eq!(batch_result, vec![records2]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index, 3)
            .unwrap();
        assert_eq!(batch_result, vec![records1, records2]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block2.index, 3)
            .unwrap();
        assert_eq!(batch_result, vec![records2]);

        // When there is a gap in the data, the gap should suppress any further results
        // even if there are hits later in the range.
        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index - 1, 2)
            .unwrap();
        assert_eq!(batch_result, vec![]);

        let batch_results = db
            .get_tx_outs_by_block_range_and_key(ingress_key, block1.index - 2, 3)
            .unwrap();
        assert_eq!(batch_result, vec![]);
```

Reply:
I'd rather not do this, I'll have to write clone everywhere I think